File: frontend/src/admin/tiptapUpload.ts
import { api } from "@admin/api";
import type { ApiResponse } from "@shared/types";

interface AdminTiptapImageUploadOutput {
  folder: string;
  path: string;
  url: string;
  content_type: string;
  size: number;
  width: number | null;
  height: number | null;
}

export async function uploadAdminTiptapImage(file: File, folder: string): Promise<{ url: string; path: string }> {
  const formData = new FormData();
  formData.append("folder", folder);
  formData.append("fileUpload", file);

  const response = await api.post<ApiResponse<AdminTiptapImageUploadOutput>>(
    "/uploads/tiptap-image",
    formData,
  );
  const payload = response.data?.data;
  if (!payload || !payload.url || !payload.path) {
    throw new Error();
  }
  return {
    url: payload.url,
    path: payload.path,
  };
}

File: app/src/contracts/api/v1/admin/tiptap_upload.rs
use schemars::JsonSchema;
use serde::Serialize;
use ts_rs::TS;

#[derive(Debug, Clone, Serialize, JsonSchema, TS)]
#[ts(export, export_to = "admin/types/")]
pub struct AdminTiptapImageUploadOutput {
    pub folder: String,
    pub path: String,
    pub url: String,
    pub content_type: String,
    #[ts(type = "number")]
    pub size: i64,
    pub width: Option<i32>,
    pub height: Option<i32>,
}
